Add a draft of joss-paper (#252)
kokkos/kokkos-fft: joss-paper/paper.md

tags:
  - C++
  - FFT
  - High performance computing
  - Performance portability
authors:
  - name: Yuuichi Asahi
    orcid: 0000-0002-9997-1274
    equal-contrib: true
    affiliation: "1" # (Multiple affiliations must be quoted)
  - name: Thomas Padioleau
    orcid: 0000-0001-5496-0013
    equal-contrib: true
    affiliation: "1" # (Multiple affiliations must be quoted)
  - name: Paul Zehner
    orcid: 0000-0002-4811-0079
    equal-contrib: true
    affiliation: "1" # (Multiple affiliations must be quoted)
  - name: Julien Bigot
    orcid: 0000-0002-0015-4304
    equal-contrib: true
    affiliation: "1" # (Multiple affiliations must be quoted)
affiliations:
 - name: Université Paris-Saclay, UVSQ, CNRS, CEA, Maison de la Simulation, 91191, Gif-sur-Yvette, France
   index: 1
   ror: 00hx57361

date: 18 April 2025
bibliography: paper.bib

# Optional fields if submitting to a AAS journal too, see this blog post:
# https://blog.joss.theoj.org/2018/12/a-new-collaboration-with-aas-publishing
aas-doi: 10.3847/xxxxx <- update this with the DOI from AAS once you know it.
aas-journal: Astrophysical Journal <- The name of the AAS journal.
---

# Summary

> kokkos-fft implements local (non-distributed) interfaces between Kokkos and de facto standard FFT libraries, including FFTW, cufft, hipfft (rocfft), and oneMKL. The key concept of kokkos-fft is "As easy as numpy, as fast as vendor libraries". This library has simple interface like numpy.fft with in-place and out-of-place transforms, while keeping the performance portability. An exercise to solve the 2D Hasegawa-Wakatani turbulence with the Fourier spectral method demonstrates that we can achieve a signficiant speed up against python version without increasing the lines of code (LOC) drastically. This library allows the user to perform FFTs efficiently and simply in the kokkos eco system.

Reviewer: ```suggestion
kokkos-fft implements local (non-distributed) interfaces between Kokkos and de facto standard FFT libraries, including FFTW, cuFFT, rocFFT, and oneMKL. The key concept of kokkos-fft is "As easy as numpy, as fast as vendor libraries". This library has simple interface like numpy.fft with in-place and out-of-place transforms, while keeping the performance portability. An exercise to solve the 2D Hasegawa-Wakatani turbulence with the Fourier spectral method demonstrates that we can achieve a signficiant speed up against python version without increasing the lines of code (LOC) drastically. This library allows the user to perform FFTs efficiently and simply in the kokkos eco system.
```
Spelling of the FFT libs. Also I would just say rocFFT and omit hipFFT for simplicity.

> kokkos-fft implements local (non-distributed) interfaces between Kokkos and de facto standard FFT libraries, including FFTW, cufft, hipfft (rocfft), and oneMKL. The key concept of kokkos-fft is "As easy as numpy, as fast as vendor libraries". This library has simple interface like numpy.fft with in-place and out-of-place transforms, while keeping the performance portability. An exercise to solve the 2D Hasegawa-Wakatani turbulence with the Fourier spectral method demonstrates that we can achieve a signficiant speed up against python version without increasing the lines of code (LOC) drastically. This library allows the user to perform FFTs efficiently and simply in the kokkos eco system.

Reviewer: ```suggestion
kokkos-fft implements local (non-distributed) interfaces between Kokkos and de facto standard FFT libraries, including FFTW, cufft, hipfft (rocfft), and oneMKL. The key concept of kokkos-fft is "As easy as numpy, as fast as vendor libraries". This library has simple interface like numpy.fft with in-place and out-of-place transforms, while keeping the performance portability. An exercise to solve the 2D Hasegawa-Wakatani turbulence with the Fourier spectral method demonstrates that we can achieve a signficiant speed up against python version without increasing the lines of code (LOC) drastically. This library allows the user to perform FFTs efficiently and simply in the Kokkos ecosystem.
```
Please cite https://doi.ieeecomputersociety.org/10.1109/MCSE.2021.3098509 for the Kokkos ecosystem

Author: Sure

Let's start with a simple example to perform the 1D real to complex transform using `rfft` in kokkos-fft.

```C++
#include <Kokkos_Core.hpp>
#include <Kokkos_Complex.hpp>
#include <Kokkos_Random.hpp>
#include <KokkosFFT.hpp>
using ExecutionSpace = Kokkos::DefaultExecutionSpace;
template <typename T> using View1D = Kokkos::View<T*, ExecutionSpace>;
int main(int argc, char* argv[]) {
  Kokkos::ScopeGuard guard(argc, argv);
  const int n = 4;
  View1D<double> x("x", n);
  View1D<Kokkos::complex<double> > x_hat("x_hat", n/2+1);
  Kokkos::Random_XorShift64_Pool<> random_pool(12345);
  Kokkos::fill_random(x, random_pool, 1);
  KokkosFFT::rfft(ExecutionSpace(), x, x_hat);
  Kokkos::fence();

> }

Author: Thanks a lot! I will update accordingly

For the `fill_random`, the argument should be range. So, it will generate 4 random elements in the range of [0, 1].
https://godbolt.org/z/aMe1x4a9Y

But, it would be nicer to use `1.0` rather than `1`.

```

This is equivalent to the following python code.

```python
import numpy as np
x = np.random.rand(4)
x_hat = np.fft.rfft(x)
```

There are two additional arguments in the Kokkos version:

* An instance of the [`ExecutionSpace`](https://kokkos.org/kokkos-core-wiki/API/core/execution_spaces.html) wherein the operations are performed.

> * The output View (`x_hat`).

Reviewer: ```suggestion
* `x_hat`: the Kokkos::View where the complex-valued FFT output will be stored. By accepting this view as an argument, the function allows the user to pre-allocate memory and optimize data placement, avoiding unnecessary allocations and copies.
```
